src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * The backoff time is between the minimum and maximum backoff time, based on the number of attempts.
+ * An exponential backoff strategy is used, with a jitter factor to prevent clients from retrying at the same time.
+ *
+ * The backoff time is calculated as follows:
+ *  - `basic backoff time` = `MinimumBackoffInMs` * 2 ^ `attempts`, and it is no larger than the `MaximumBackoffInMs`.
+ *  - based on jitter ratio, the jittered time is between [-1, 1) * `JitterRatio` * basic backoff time.
+ *  - the final backoff time is the basic backoff time plus the jittered time.
+ *
+ * Note: the backoff time usually is no larger than the refresh interval, which is specified by the user.
+ *  - If the interval is less than the minimum backoff, the interval is used.
+ *  - If the interval is between the minimum and maximum backoff, the interval is used as the maximum backoff.
+ *  - Because of the jitter, the maximum backoff time is actually `MaximumBackoffInMs` * (1 + `JitterRatio`).
+ */
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxSafeExponential = 30; // Used to avoid overflow. bitwise operations in JavaScript are limited to 32 bits. It overflows at 2^31 - 1.
+const JitterRatio = 0.25;
+
+export class RefreshTimer {

```

Comment:
Is there a way to make this class internal such that our provider is the only consumer of its methods?

Reply:
Currently it is kind of internal, because we don't export it in `src/index.ts`.  Here `export` means you can reference it in other files/modules.
For users using CJS module system, it's internal, no way to import "RefreshTimer" in user code. 
For users using ESM module system, they can import RefreshTimer from "path-to-refreshTimer.ts" if they insists.

If you want to ensure it's 100% internal, we can move it into the file using it, and remove `export`. But that's not a big issue. I would rather keep it as a single file for readability...

Reply:
Sounds good, lets keep it separate